Makefile.Common

```diff
@@ -11,14 +11,15 @@ IMPI=impi
 GOOS := $(shell $(GOCMD) env GOOS)
 GOARCH := $(shell $(GOCMD) env GOARCH)
 GH := $(shell which gh)
+COVERAGE_PKGS=$(shell go list ./... | grep -v testutils)
```

Comment:
Not sure I agree with `testutils` being ignored here, I understand it is a test package and testing tests feels redundant however, they do server as excellent examples for those important and using the package.

Furthermore, it helps check if things change, would the interactions with the library change?

Reply:
When you mean things change as in things like interface change, etc? All the changes should really be reflected on the test that uses the library. So if things change, those tests should fail and if they don't the tests were written properly.

Unless this is a widely used test library across a lot of packages internally/externally I don't think it really makes sense to have tests on them.

Reply:
I also don't see a reason to exclude it. If the package contains test helpers, it means that it must be 100% covered. Otherwise, some helpers are not being used

Reply:
`MockStorageClient` implements the interface of `Client` and there are 5 methods inside the interface. It is possible the tests are not using all methods but implementation for all methods are required.

Reply:
I believe it's ok to keep them uncovered if they are noop and added just to implement the interface

Reply:
I tried it again and the the method is still flagged as not covered.

I notice another package `internal/testutil` with some other testutil codes that have unit test coverage. I will just add some basic test to make this easier. Also renaming the struct from `MockStorageClient` to `InMemoryStorageClient`

Reply:
> Also renaming the struct from MockStorageClient to InMemoryStorageClient

Why? `InMemoryStorageClient` sounds like a production-use client. I think `MockStorageClient` is a better name if it's intended to be used in tests

Reply:
`MockStorageClient` isn't explicit on what the storage client is doing. The word `Mock` can confuse whoever is using it. For example, if they have to do specific setups on the Mock? Do they have to mock input/output like the `mockery` library?

`InMemoryStorageClient` represents exactly what the struct is doing and being in a commonly used shared package it is important to have names that people can understand immediately. The package name is `testutil` and it is under an internal package so there are some safe guards to prevent people to mistake it as a production-use client.

Reply:
Why do we need to mock persistent storage with in-memory storage if in-memory storage is the default anyway? Can we just don't set up the persistent storage in this test? It looks like the only side effect we need here is setting `requeuingEnabled` on `queuedRetrySender` which I believe can be set right in the test. In that case, we don't need the `testutil` package.

Reply:
The default persistent storage is the file storage client but the test is using in memory storage client. I am not sure why the initial setup was this way but my guess is the test runs don't necessarily have correct files permission access?

`queued_retry` does initialize [storage](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L165) so we have to set some kinds of storage to test the scenarios. I think using the persistent_queue directly can tests these data loss scenarios nicely but it requires a storage client.
Another alternative is to write some mock queue so we can over write the queue in retry_sender. This way it is more unit test like but at the same time we can miss testing these functionalities all together.

Reply:
This is what i mean by the alternative. The test is written directly against the retrySender. Then we don't have to do anything with the storageClient. However, I was hoping we can have some kinds of tests that involve multiple components so we know data loss can't happen with persistent queue. Maybe there can be some kinds of functional tests instead of unit tests.

```
func TestRetrySender_shutdown_tempError(t *testing.T) {
	req := newMockRequest(context.Background(), 3, errors.New("some error"))

	qCfg := NewDefaultQueueSettings()
	qCfg.NumConsumers = 1
	rCfg := NewDefaultRetrySettings()
	rCfg.InitialInterval = time.Hour // set high initial retry interval so it only waits
	rCfg.MaxElapsedTime = 0          // retry infinitely so shutdown can be triggered

	be, err := newBaseExporter(defaultSettings, fromOptions(WithRetry(rCfg), WithQueue(qCfg)), "", nopRequestUnmarshaler())
	require.NoError(t, err)
	require.NoError(t, be.Start(context.Background(), &mockHost{}))

	// replace nextSender inside retrySender to always return error so it doesn't exit send loop
	castedSender, ok := be.qrSender.consumerSender.(*retrySender)
	require.True(t, ok, "consumerSender should be a retrySender type")
	castedSender.nextSender = &ErrorRequestSender{
		errToReturn: errors.New("some error"),
	}
	onTempFailureErrorChan := make(chan error, 1)
	castedSender.onTemporaryFailure = func(_ *zap.Logger, _ internal.Request, err error) error {
		onTempFailureErrorChan <- err
		return nil
	}

	sendErrChan := make(chan error, 1)
	go func() {
		sendErrChan <- be.sender.send(req)
	}()
	require.NoError(t, <-sendErrChan)

	// shuts down so onTemporaryFailure is invoked
	require.NoError(t, be.Shutdown(context.Background()))
	require.ErrorContains(t, <-onTempFailureErrorChan, "interrupted due to shutdown")
}
```

Reply:
Can't you use your original test, but instead of:

```golang
	storageID := component.NewIDWithName("file_storage", "storage")

	qCfg := NewDefaultQueueSettings()
	qCfg.NumConsumers = 1
	qCfg.StorageID = &storageID // enable persistence
	rCfg := NewDefaultRetrySettings()
	rCfg.InitialInterval = time.Millisecond
	rCfg.MaxElapsedTime = 0 // retry infinitely so shutdown can be triggered

	be, err := newBaseExporter(defaultSettings, fromOptions(WithRetry(rCfg), WithQueue(qCfg)), "", nopRequestUnmarshaler())
	require.NoError(t, err)

	var extensions = map[component.ID]component.Component{
		storageID: &mockStorageExtension{
			mockClient: testutil.NewInMemoryStorageClient(),
		},
	}
	host := &mockHost{ext: extensions}

	// we start correctly with a file storage extension
	require.NoError(t, be.Start(context.Background(), host))
```

do this:

```
	qCfg := NewDefaultQueueSettings()
	qCfg.NumConsumers = 1
	rCfg := NewDefaultRetrySettings()
	rCfg.InitialInterval = time.Millisecond
	rCfg.MaxElapsedTime = 0 // retry infinitely so shutdown can be triggered

	be, err := newBaseExporter(defaultSettings, fromOptions(WithRetry(rCfg), WithQueue(qCfg)), "", nopRequestUnmarshaler())
	require.NoError(t, err)
	be.qrSender.requeuingEnabled = true

	require.NoError(t, be.Start(context.Background(), componenttest.NewNopHost()))
```

Reply:
Yeah actually this works. I can make the change